src/charm.py

```diff
@@ -80,11 +85,45 @@ def _on_config_changed(self, _):
         else:
             self.unit.status = WaitingStatus("Waiting for Pebble in workload container")
 
+    def _patch_auth_resources(self):
+        """Returns True if existing auth resources for training-operator are patched correctly."""
+        self.unit.status = MaintenanceStatus("Patching auth configuration")
+        client = Client()
+        context = {"namespace": self._namespace}
+
+        with open(os.path.join(self._src_dir, "auth_manifests.yaml")) as f:
+            for obj in codecs.load_all_yaml(f, context=context):
+                try:
+                    client.patch(
+                        type(obj), obj.metadata.name, obj, patch_type=PatchType.MERGE
+                    )
+                except ApiError as e:
+                    logging.error(f"Patching auth resources failed: {str(e)}")
+        logging.info("Auth resources successfully patched.")
+        return True
+
+    def _create_crds(self):
+        """Returns True if CRDs for training-operator are created correctly."""
+        self.unit.status = MaintenanceStatus("Creating CRDs")
+        client = Client()
+
+        with open(os.path.join(self._src_dir, "crds_manifests.yaml")) as f:
+            for obj in codecs.load_all_yaml(f):
+                try:
+                    client.create(obj)
+                except ApiError as e:
+                    logging.error(f"Creating CRDs failed: {str(e)}")
+        logging.info("CRDs successfully created.")
+        return True
 
     def _on_install(self, _):
-        """Placeholder for install event"""
-        pass
+        """Event handler for on-install events."""
+        auth = self._patch_auth_resources()
+        crds = self._create_crds()
 
+        # Check resources were applied/created correctly
+        if auth and crds:
+            self.unit.status = ActiveStatus()
```

Comment:
What if these have not successfully applied? What state do you want to leave the charm in?

Once the above methods are changed, one of these conditions could be `False`, which I think would leave the charm in `MaintenanceStatus`. I'm not actually sure what the correct status is here; you could put the charm into `BlockedStatus`, and then provide an action that tries to re-apply the manifests? 

Probably worth some discussion - in some cases (like in the K8s service patcher library), I just log and move on, because the Service patching doesn't *stop* the application from running. In the case that something is required for an app to even function, Blocked seems appropriate, but in the case that it's not critical, ActiveStatus with a message might be a good idea?

Reply:
The case when CRDs already exist on the cluster and no further action is needed is covered in `_crds_creation`. I think that one is not critical, so returning `True` there and setting `ActiveStatus` later in the event handler should be okay. 
For cases when resources were not created/patched correctly, +1 on using `BlockedStatus`. By definition it is a status that will inform users when something went wrong with config, and manual intervention is needed, though I am not sure about the manual intervention part, shall we just provide the exception and let users sort out the rest?

Reply:
Seems like a particularly interesting course to pursue in the context of how most of the KF charms work broadly outside of Juju's control.

Reply:
Perhaps we don't need an extra `validate_resource` action as we are already iterating over the resources within `_create_crds()` and `_patch_auth_resources()`, and when an exception occurs, we know exactly what resource is having the issue as well as the failure reason, and other interesting and useful info. 

Regarding the action users should do in the event of failure, I like this idea here:
> ...checks if a corresponding resource exists on the cluster, and dumps the manifest if not?

That being said, we can set `BlockedStatus`, and dump the exception msg and manifest of the missing resource (we can even save the yaml file in a tmp dir) with instructions on how to create/patch the resource after the issue that caused the exception is resolved.

That, or a mechanism by which users can tell juju they've resolved the issue and let it re-try to patch/create resources again, without needing humans to manage any manifest files. Though for this I wouldn't know how to do it, perhaps in the config-changed event handler.

Reply:
I agree with @DnPlas that for now, we should set `BlockedStatus`, and log the resource that failed to create along with the traceback. That will be enough for an admin to come in and know what's wrong. I like where @jnsgruk and @ca-scribner are going with this in general, but I think a lot of that work is something that Juju should negotiate with Kubernetes, and not something a charm author should have to write themselves.

For example, I don't think every K8s charm should be copy/pasting a `validate_resource` action, that's the sort of thing Juju itself should provide. Particularly with Juju Templates, we could tell Juju "create these resources", and if any of them failed to create, Juju would automatically make that apparent in `juju status`. We could then have some sort of `juju show-failed-resources` command to make it easily visible what failed to create and why.

Reply:
I agree with @knkski that Juju should provide this kind of functionality. I think k8s native charms will be benefited by something like `juju validate-resources` and/or `juju show-failed-resources`. I'd like to see something like that, though extending juju's capabilities are out of the scope of this PR :) 
For now, I'm going to submit a new version of this PR letting the helper methods show all useful data, and the event handler set the status. As always, I'd appreciate your comments.

Reply:
One extra thing. I would like to know your thoughts on using `juju resolved`. If we provide enough info about what is causing the issue, can we go with something like `BlockedStatus("Resolve <issues> and run juju resolved afterwards)`?

Reply:
I feel like logging the failed resource and asking the user to run juju resolved afterwards is the right approach. Please add these changes and let's get this PR merged.